### PR TITLE
Resize 3D + resize tests

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature: 3D resize, needed in medical imaging and more.
- It adds missing high-level tests for Resize (2D and 3D)

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Parameterized the kernel with spatial_ndim
     * Extend value switch to handle 3 dims
     * Add extra new layouts to the list of supported ones
     * Remove/loosen constraints
     * Extensive tests against two-step resize with PIL and numpy
 - Affected modules and functionalities:
     * Resize op
 - Key points relevant for the review:
     * Tests!
 - Validation and testing:
     * Extensive Python-level tests for resize op
 - Documentation (including examples):
     * Updated resize example
     * Updated operator documentation
     * Updated 3D transforms example (now uses MRI data and resize is a pre-step).

**JIRA TASK**: DALI-1076

### CODE REVIEW

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/operators/image/resize/resize_attr.cc [hunk 1]
```diff
@@ -64,7 +64,13 @@ This option is equivalent to specifying the same size for all dimensions and ``m
   .AddOptionalArg<vector<float>>("max_size", R"(Limit of the output size - when resizing using
 ``resize_shorter``, "not_smaller" mode or otherwise leaving some extents unspecified, some images
 with high aspect ratios might produce extremely large outputs. This parameter puts a limit to how
-big the output can become. This value can be specified per-axis or uniformly for all axes.)",
+big the output can become. This value can be specified per-axis or uniformly for all axes.
+
+NOTE: When used with "not_smaller" mode or ``resize_shorter`` argument, ``max_size`` takes
```
Comment (JanuszL) [inline dali/operators/image/resize/resize_attr.cc:69]:
```suggestion
.. note::
  When used with "not_smaller" mode or ``resize_shorter`` argument, ``max_size`` takes
```

File: dali/operators/image/resize/resize_attr.cc [hunk 2]
```diff
@@ -64,7 +64,13 @@ This option is equivalent to specifying the same size for all dimensions and ``m
...
+precedence and the aspect ratio is kept - for example, resizing with
+``mode="not_smaller", size=800, max_size=1400`` an image of size 1200x600 would be resized to
+1400x700.
```
Comment (JanuszL) [inline dali/operators/image/resize/resize_attr.cc:72]:
```suggestion
.. note::
  When used with "not_smaller" mode or ``resize_shorter`` argument, ``max_size`` takes
  precedence and the aspect ratio is kept - for example, resizing with
  ``mode="not_smaller", size=800, max_size=1400`` an image of size 1200x600 would be resized to
  1400x700.
```

File: docs/examples/image_processing/resize.ipynb [hunk 3]
```diff
@@ -8,11 +8,13 @@
     "\n",
     "This notebook demonstrates the usage of Resize operator.\n",
     "\n",
-    "Resize operator resamples images and sequences of images. It can be parameterized with respect to how the final output size is calculated and what kind of filtering or interpolation method is applied. It also supports resizing regions of interest (RoI).\n",
+    "Resize operator resamples images, volumes* and sequences of images (and volumes). It can be parameterized with respect to how the final output size is calculated and what kind of filtering or interpolation method is applied. It also supports resizing regions of interest (RoI).\n",
```
Comment (JanuszL) [inline docs/examples/image_processing/resize.ipynb:11]: Volumes and ... (and volumes) ?

File: docs/examples/image_processing/resize.ipynb [hunk 4]
```diff
@@ -8,11 +8,13 @@
...
     "\n",
     "> **Note** `DALI_EXTRA_PATH` environment variable should point to the place where data from [DALI extra repository](https://github.com/NVIDIA/DALI_extra) is downloaded. Please make sure that the proper release tag is checked out.\n",
     "\n",
-    "> **Note** For detailed description of filtering methods, see \"Interpolation methods\" tutorial."
+    "> **Note** For detailed description of filtering methods, see \"Interpolation methods\" tutorial.\n",
```
Comment (JanuszL) [inline docs/examples/image_processing/resize.ipynb:15]: We can't link it here, right?

File: dali/test/python/test_operator_resize.py [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+
+        dali_interior = [interior(x, channel_first) for x in dali_resized]
+        ref_interior =  [interior(x, channel_first) for x in ref_resized]
+        check_batch(dali_interior, ref_interior, batch_size, max_avg_err, max_err)
+
+def _tests(dim, device):
+    batch_size = 2 if dim == 3 else 10
+    # 1. Cannot test linear against PIL, because PIL uses triangular filter when downscaling
+    # 2. Cannot test Nearest Neighbor because rounding errors cause gross discrepancies (pixel shift)
+    for mode in ["default", "stretch", "not_smaller", "not_larger"]:
```
Comment (JanuszL) [inline dali/test/python/test_operator_resize.py:421]: Can't we implement NN using numpy?

Comment (mzient) [reply dali/test/python/test_operator_resize.py:421]: It's not a question if we can or not - it's rather that it's hard to compare. The differences in NN can cause almost unbounded changes in the output (you can jump from black pixel to the neighboring white pixel). For tests I'd have to implement a fused resize-and-test function which would be rather involved.

Comment (JanuszL) [reply dali/test/python/test_operator_resize.py:421]: ok

File: docs/examples/image_processing/resize.ipynb (hunk 4 above)
Comment (mzient) [reply docs/examples/image_processing/resize.ipynb:15]: I don't know, honestly.

**Review by JanuszL — APPROVED** [review]

**Comment by mzient** [comment]
!build

**Review by awolant — APPROVED** [review]

**Pull request merged**